src/main/scala/chisel3/internal/firrtl/Emitter.scala

```diff
@@ -82,7 +82,7 @@ private class Emitter(circuit: Circuit) {
     */
   private def emit(m: Component): String = {
     // Generate the body.
-    val moduleName = m.id.getClass.getName.split('.').last
+    val moduleName = m.id.desiredName
```

Comment:
I need to do an internal cleanup first, but I will try to upstream our fix for this issue.

(Was okay with waiting when your PR was just going to go through but now there isn't much sense in having you continue to duplicate developing the same annoying edge case resolutions.)